client/state/add-reducer.ts

```diff
@@ -26,9 +26,9 @@ async function initializeState(
 // For a given store, creates a function that adds a new reducer to the store,
 // and loads (asynchronously) and applies the persisted state for it.
 export const addReducerToStore = ( store: object ) => (
-	key: string,
+	key: string[],
```

Comment:
why do we pass in the key as an array instead of a single string? I'm pretty sure that if we passed in `'a.b.c'` instead of `['a','b','c']` then we could track the keys in the type system though not with `Map()`.

![Screen Shot 2019-05-07 at 9 54 23 AM](https://user-images.githubusercontent.com/5431237/57317961-1fb3ac80-70ae-11e9-9967-7643eeb45c4d.png)
[Try it out](https://www.typescriptlang.org/play/index.html#src=const%20a%3A%20%7B%20%5Bkey%3A%20string%5D%3A%20boolean%20%7D%20%3D%20Object.create(null)%3B%0D%0A%0D%0Afunction%20add%3C%0D%0A%20%20%20%20T%20extends%20%7B%20%5Bkey%3A%20string%5D%3A%20boolean%20%7D%2C%0D%0A%20%20%20%20K%20extends%20keyof%20T%0D%0A%3E(%0D%0A%20%20%20%20o%3A%20T%2C%0D%0A%20%20%20%20key%3A%20K%2C%0D%0A%20%20%20%20value%3A%20boolean%0D%0A)%3A%20T%20%26%20%7B%20%5BKey%20in%20K%5D%3A%20boolean%20%7D%20%7B%0D%0A%20%20%20%20return%20Object.assign(%0D%0A%20%20%20%20%20%20%20%20Object.create(null)%2C%0D%0A%20%20%20%20%20%20%20%20o%2C%0D%0A%20%20%20%20%20%20%20%20%7B%20%5Bkey%5D%3A%20value%20%7D%0D%0A%20%20%20%20)%3B%0D%0A%7D%0D%0A%0D%0Aconst%20b%20%3D%20add(a%2C%20'empty'%2C%20false)%3B%0D%0Aconst%20c%20%3D%20add(b%2C%20'fun'%2C%20true)%3B%0D%0A%0D%0Ac.)

Reply:
@jsnajdr may know more, but I'd venture a guess that it's a way of avoiding string parsing like in `_.get`.

Reply:
> I'd venture a guess that it's a way of avoiding string parsing like in _.get

that was my first thought but I didn't find any string parsing, just `keys.join( '.' )`

also, is this stuff in anticipation things to come? I didn't find any modules at a glance that seemed to dynamically add reducers. I did a cursory search for default exports in `index.js` files and found some that use the `router` argument but none that took the `addReducerToStore` function

Reply:
> why do we pass in the key as an array instead of a single string?

I don't fully understand the question. The object keys are strings, and the key chain for a nested reducers is therefore a list of strings. The array is an obvious choice. Passing a dotted string would be a convenient shortcut, but it's not really needed.

How would an `a.b.c` string help derive the return type of `addReducer`? Even with the array, it should be possible to define the type recursively:
- if key matches `[ k ]`, then the type of the new state is `OriginalState & { [ k ]: AddedState }`
- if key matches `[ k, ...restK ]`, then the type of the new state is `OriginalState & { [ k ]: typeof addReducer( restK )`
(hope my pseudocode makes sense 🙂)

But even if we are able to precisely define the return type of `addReducer`, I don't see how it would be practically useful. Redux state is a global variable, and what we'd need is to change the type of that global variable on each `addReducer` call. Because the global variable is what all the components use:
```js
connect( state => { ... } )
```
The type of `state` should change whenever someone calls `addReducer`. For N possible dynamic reducers, there are 2<sup>N</sup> possible types `state` can have. It all doesn't look feasible at all.

> also, is this stuff in anticipation things to come? I didn't find any modules at a glance that seemed to dynamically add reducers.

All extensions have dynamic reducers, most important of them being WooCommerce. Another section that uses it is `me/concierge`. Signup, Login and Reader are likely to follow soon. Unfortunately, all of them have one or two weird use cases where the reducer is used (or misused) outside the section that owns it.

Reply:
> The object keys are strings, and the key chain for a nested reducers is therefore a list of strings.

While looking at it I thought I was seeing that we are using the dot-joined string as the key for the `Map()`, in which case it doesn't seem like they are nested properties but just a convenient way to organize the reducer data.

> How would an a.b.c string help derive the return type of addReducer?

The primary difference is that `'a.b.c'` is a string literal which can have its own type vs. joining an array of strings which loses the actual names. So, if we pass in `[ 'a', 'b', 'c' ]` and want to track it then if my understanding is correct we'd have to recreate a string join in the type system and that would be pretty insane for anyone to read (it can be done).

With a string literal key then something like my example above demonstrates how we can track those dynamic additions through the code after they have been added. I'm not sure how far we could take that, and as you point out, currently that affects a global value which makes the types even weirder.

I'm happy leaving things as they are - just wondering why the array since when working with types it stood out to me.

> For N possible dynamic reducers, there are 2<sup>N</sup> possible types state can have.

You lost me 😄 . Where did you get 2<sup>N</sup>?

My dream would be to have the entire state tree typed so that we could write `state.sites[ … ].…` and have it fully auto-complete and warn us when we are accessing something that's not there. I don't think that's infeasible so much as not possible given our current code. I think that's a later-term project for the future too (vs. right now).

Reply:
> also, is this stuff in anticipation things to come? I didn't find any modules at a glance that seemed to dynamically add reducers. I did a cursory search for default exports in `index.js` files and found some that use the `router` argument but none that took the `addReducerToStore` function

`addReducerToStore` isn't called directly by any section. Instead, it's passed to a section's main function as the second argument, through the section middleware. Sections making use of this include various extensions and concierge.

```js
export default async ( _, addReducer ) => {
	await addReducer( [ 'concierge' ], reducer );
       ...
};
```

Reply:
Also, I want to emphasise that this file isn't self-contained. The keys it receives aren't used internally only; they're passed on to `store.addReducer()`, which is the one that expects them to be an array of strings. For this particular file, it would be best if they *weren't* an array of strings, which is why I convert them to a dot-separated string with `normalizeKey()`.

Hope this helps!

Reply:
it was the `async` part that threw off my search so I didn't notice the extensions code (@jsnajdr also pointed out that extensions use them)

thanks for clarifying!

Reply:
> Where did you get 2N?

Every of the N dynamic reducers either is or is not there. That's one bit for each one. Depending on which ones got loaded or not, the number of possible state types is the largest N-bit number 🙂